internal/sync-pnpm/index.js

```diff
@@ -30,33 +33,81 @@ export default async function syncPnpm(dir = process.cwd()) {
       ownDependencies.includes(p.packageJson.name)
   );
 
+  let paths = {};
+
   for (const pkg of packagesToSync) {
     const syncFrom = join(pkg.dir, syncDir);
+
     const resolvedPackagePath = dirname(
       resolvePackagePath(pkg.packageJson.name, dir)
     );
     const syncTo = join(resolvedPackagePath, syncDir);
 
     if (await pathExists(syncFrom)) {
-      let releaseLock;
-      try {
-        releaseLock = await lockfile.lock(syncTo, { realpath: false });
-        debug(`lockfile created for syncing to ${syncTo}`);
-      } catch (e) {
-        debug(
-          `lockfile already exists for syncing to ${syncTo}, some other sync process is already handling this directory, so skipping...`
-        );
-        continue;
-      }
+      paths[syncFrom] = syncTo;
+    }
+  }
+
+  if (!watchMode) {
+    for (const [syncFrom, syncTo] of Object.entries(paths)) {
+      syncDependency(syncFrom, syncTo);
+    }
+
+    return;
+  }
+
+  let fromPaths = Object.keys(paths);
+  let watcher = new Watcher(fromPaths);
+
+  let dirtyPaths = [];
 
-      if (await pathExists(syncTo)) {
-        await remove(syncTo);
-        debug(`removed ${syncTo} before syncing`);
+  watcher.on('all', (event, targetPath, targetPathNext) => {
+    dirtyPaths.push(targetPath);
```

Comment:
`dirtyPaths` contains a list of all of the files that have changed but we need to reduce that to a list of the _packages_ that have changed (e.g. entries in `fromPaths`)... Because the sync hard links the whole package directory, not individual files within it.

I did see the `debounce` option but this bit of the docs:

> the higher this is the more duplicate events will be ignored automatically.

Made me think it's not super relevant... I assume a "duplicate" event would be changes to a single file but we are interested in changes within a single _package_.

Doing the debouncing ourselves here seemed trivial and allowed me to understand what was happening and `log` the different places while debugging. I don't think any performance overhead would be critical and with a `DEBOUNCE_INTERVAL` of `50` everything seems instant enough to me when working on it..

Reply:
> Because the sync hard links the whole package directory, not individual files within it.

Yeah, ok, but that came from the times where we explicitly called the sync script to do all syncing as a one-off job, as part of turbo or manually. We could totally have granular syncs now also, right? Not sure if it's worth optimizing this at this point though...

What about my 2nd paragraph above (about the debounce timing), don't you think that is relevant?

Reply:
> We could totally have granular syncs now also, right? 

I guess... I wasn't sure if there might be interdependencies between files or if we would end up triggering multiple ember builds...

> What about my 2nd paragraph above (about the debounce timing), don't you think that is relevant?

I guess it's theoretically possible but I haven't seen any practical consequence of it. *If* it does happen then it should right itself within 50ms, shouldn't it?

Reply:
> I guess it's theoretically possible but I haven't seen any practical consequence of it. If it does happen then it should right itself within 50ms, shouldn't it?

With the recent changes I did see a failing app build once, but I only noticed this by carefully looking at the log output. The app rebuild instantly after that, so that was not an issue.

I guess we could try to optimize this, e.g. by tweaking our debounce logic. But I think we can leave this for future iterations...